### ISSUE TITLE
Feature: binding string interpolation

### ISSUE THREAD

**Issue by kevinpschaaf** [Polymer/polymer#2182]
Support for multiple annotated bindings in properties, attributes, and textContent.

Should support the following use cases:

```html
<my-el class$="foo bar {{baz}} zot {{computeClass(a, b)}} ziz">
  My name is: {{name}} <br>
  I like to {{computeLikes(a, b)}}!!
</my-el>
```

**Comment by robrez** [comment]
:+1:

**Comment by eximius313** [comment]
+1

**Comment by Jamtis** [comment]
Logic belongs in the script!
But how complex must a expression be to be logic?
Is the example so complex that you should write a computing method for the whole `class` attribute?

To me it seems a bit to much effort to make up an own computing method for the example given.

But I want to put in consideration:

How powerful must/will the parser be?

I think it's important to limit the competence of the parser to simply replacing bracket-expressions and not allow (complex) syntax in the markup.

That could lead to things like the old times:
`<div onclick="no.syntax=false;event.dispatch(this);">`

So, limiting the power of the parser will keep the core nice and tight and the logic separated from the content.

Finally, +1

**Comment by eximius313** [comment]
but how about this example:
```<div class="{{model.promoted?'promoted ':''}}{{model.id == activeId?'active':''}}">``` 
where the "computation" is very simple BUT activeId can change any time, so it must be reflected immediately (this is how it works in Polymer 0.5)

**Comment by Jamtis** [comment]
IMO, that's logic.
So it should be in the script/method.

That is a conditional expression which is a logic element.

**Comment by eximius313** [comment]
1)So you'll write separate method in JS, just to check one boolean statement?
2)How the change to ```activeId``` will be reflected back in div's class if I compute it just once?

**Comment by Jamtis** [comment]
Yes.

Like this:

`<div class$="{{_computeClass(model.promoted, model.id, activeId)">`

and

```
_computeClass(promoted, id, activeId) {
return (promoted ? "promoted" : "") + (activeId == id ? "active" : "");
}
```

**Comment by eximius313** [comment]
1)IMHO, this approach will blow up HTML and JS file
2)Correct me if I'm wrong, but once computation is done, it will not reflect changes of ```activeId```, right?

**Comment by Jamtis** [comment]
I think "blow up" is a bit extreme.

I give more credit to the performance of the polymer core, in regard of the necessary parser, than the file size of a certain component.

Secondly a recomputation is done whenever a argument changes, including `activeId`.

**Comment by eximius313** [comment]
1)I've already done it in real life app and I have dozens of "return only" functions
2)I have tested it just second ago and It didn't work

**Comment by Jamtis** [comment]
I'd appreciate you create a fiddle or something, so I can check on that.

**Comment by eximius313** [comment]
Which of JSFiddle/Bin/else supports Polymer?
So far it throws ```"compute method `_computeClass` not defined"```... https://jsbin.com/yuqisexebo/1/edit?html,css,console,output

And back to my code - I think that the problem occurs with ```<template is="dom-repeat" items="{{items}}">```. It looks like ```class="{{_compute(something)}}"``` is completely ignored.

**Comment by Jamtis** [comment]
The occuring error is caused by the element recognizing the changed attribute value to `{{_computeClass(..)}}` and not finding the (not yet implemented) `_computeClass` method.

I guess in `dom-module` elements it works because the method is implemented before the value is actually computed.

The current error is due to a lack of support for template computed methods and not due to the incompetence of computed methods to update on property changes.

I would like to open another issue on that specific topic with your example.

**Comment by eximius313** [comment]
ok, do it please.
And is the explosion of "return only functions" still under concern of this issue?

**Comment by Jamtis** [comment]
I can across [this issue](2237) which should cover your concerns.

I hope with behaviors we can effectively tackle both problems to anyones convenience.

**Comment by Jamtis** [comment]
PS:
The "class" problem is due to the assignment operator you use.
Use `$=` instead of `=`.
https://jsbin.com/namoxiketi/1/edit?html,css,console,output

**Comment by eximius313** [comment]
ok, ```$=``` solved the problem of recomputation. Thank you @Jamtis 
However having this: ```<my-el class$="foo bar {{baz}} zot {{computeClass(a, b)}} ziz">``` instead of dozens of ```compute``` methods will be great!

**Comment by ebidel** [comment]
@kevinpschaaf how will this affect performance? Wasn't interpolation a source of perf concern in 0.5? Or was it just the arbitrary expression syntax?

**Comment by nazar-pc** [comment]
I guess string interpolation might be eventually represented as computed binding which is already present, this is just syntactic sugar that will add some performance hit at initial parse time, but will not differ from other computed bindings in runtime.
Namely:
```html
<my-el class$="foo bar {{baz}} zot {{computeClass(a, b)}} ziz">
  My name is: {{name}} <br>
  I like to {{computeLikes(a, b)}}!!
</my-el>
```
Will be equivalent to:
```html
<my-el class$="[[something('foo bar ', baz, ' zot ', a, b, ' ziz')]]">
  [[details(' My name is: ', name, ' <br> I like to ', a, b, '!!')]]
</my-el>
```

While `something` and `details` will be generated at parse time.

**Comment by HaykoKoryun** [comment]
Since we already have multiple syntaxes, namely `{{ }}` and `[[ ]]` which do slightly different things, why not use the underscore from `0.5` to denote that the attribute contains multiple bindings, or needs the strings to be concatenated and thus should be treated differently. This way if you don't use `class_=` in your markup, then you don't experience any performance issues.

Right now, for something simple as changing the state of an indicator and displaying a hover effect, I have to write this:

`class$="{{concat('icon large', level, state)}}"` with a `concat` function where it could have been as simple as `class_="icon large {{level}} {{state}}"`

**Comment by Jamtis** [comment]
If I understood @nazar-pc right then concatenation syntax doesn't affect runtime perf.
So I don't see a need for an underscore operator.

The original perf concerns I raised was about arbitrary expression syntax, like the ternary operator in expressions.

One could suggest that we use the underscore for arbitrary syntax.
But I wouldn't do that either, for two reasons:
1. Logical expressions shouldn't be in the semantics.
2. Developers had to deal with to different operators. It allowed devs to be lazy and always use the underscore with a downside on perf.

**Comment by tomalec** [comment]
:+1:

**Comment by mgiuffrida** [comment]
:+1:

**Comment by dsschnau** [comment]
:+1:
